server/conn.go

```diff
@@ -210,6 +214,35 @@ func (cc *clientConn) readPacket() ([]byte, error) {
 	return cc.pkt.readPacket()
 }
 
+func (cc *clientConn) readPacketWithTimeOut() ([]byte, error) {
+	valStr, _ := cc.ctx.GetSessionVars().GetSystemVar(variable.WaitTimeout)
+	waitTimeout, err := strconv.ParseUint(valStr, 10, 64)
+	if err != nil {
+		log.Errorf("con:%d get sysval wait_timeout error.", cc.connectionID)
+		// if get waitTimeout error, use default value
+		waitTimeout = 28800
+	}
+
+	chanPacketByte := make(chan []byte, 1)
+	chanPacketErr := make(chan error, 1)
+	go func() {
+		data, err := cc.readPacket()
+		chanPacketErr <- err
+		chanPacketByte <- data
+		close(chanPacketByte)
+		close(chanPacketErr)
+	}()
+	select {
+	case <-time.After(time.Duration(waitTimeout) * time.Second):
```

Comment:
how about using https://golang.org/pkg/net/#TCPConn.SetReadDeadline instead of timer~?

Reply:
SetReadDeadline is better, I will refactor.